### PR TITLE
fixes #20149; fixes #16762; hintAsError and warningAsError now ignore foreign packages

### LINKED ISSUE #16762: warningaserror doesn't respect package boundaries (foreign package notes)
### Example 1
```nim
import os
```

### Current Output
XDG_CONFIG_HOME= nim r --warningaserror:deprecated main
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(143, 30) Error: NimIdent is deprecated [Deprecated]

### Expected Output
works because macros is a foreign package wrt main

### Example 2
```nim
import os
echo existsDir(".")
```

XDG_CONFIG_HOME= nim r --warningaserror:deprecated main2

the relevant error appears only with `--errormax:50`

/Users/timothee/git_clone/nim/timn/tests/nim/all/t11723.nim(6, 6) Error: use dirExists; existsDir is deprecated [Deprecated]

and there are some strange errors, eg:

times.nim(2019, 28) Error: can raise an unlisted exception: ref TimeFormatParseError

```
XDG_CONFIG_HOME= nim r --warningaserror:deprecated --errormax:50 --skipparentcfg $timn_D/tests/nim/all/t11723.nim
Hint: used config file '/Users/timothee/git_clone/nim/Nim_devel/config/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/git_clone/nim/Nim_devel/config/config.nims' [Conf]
..........
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(143, 30) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(147, 18) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(155, 18) Error: NimSym is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(234, 26) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(237, 27) Error: NimSym is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(240, 18) Error: NimSym is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(252, 16) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(256, 16) Error: NimSym is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(387, 34) Error: NimSym is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(390, 33) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(450, 23) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(453, 16) Error: Deprecated since version 0.18.1; Generate a new 'NimNode' with 'ident(string)' instead.; ident= is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(687, 24) Error: NimIdent is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(692, 14) Error: newIdentNode is deprecated [Deprecated]
........
/Users/timothee/git_clone/nim/Nim_devel/lib/posix/posix.nim(879, 23) Error: use `csize_t` instead; csize is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/posix/posix.nim(879, 31) Error: use `csize_t` instead; csize is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/posix/posix.nim(885, 21) Error: use `csize_t` instead; csize is deprecated [Deprecated]
/Users/timothee/git_clone/nim/Nim_devel/lib/posix/posix.nim(885, 29) Error: use `csize_t` instead; csize is deprecated [Deprecated]
...
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2024, 20) template/generic instantiation of `format` from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2017, 65) template/generic instantiation from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2019, 28) Error: can raise an unlisted exception: ref TimeFormatParseError
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2038, 20) template/generic instantiation of `format` from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2017, 65) template/generic instantiation from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2019, 28) Error: can raise an unlisted exception: ref TimeFormatParseError
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2125, 8) template/generic instantiation of `parse` from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2110, 22) template/generic instantiation from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2112, 28) Error: can raise an unlisted exception: ref TimeFormatParseError
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2143, 20) template/generic instantiation of `format` from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2017, 65) template/generic instantiation from here
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/times.nim(2019, 28) Error: can raise an unlisted exception: ref TimeFormatParseError
/Users/timothee/git_clone/nim/timn/tests/nim/all/t11723.nim(6, 6) Error: use dirExists; existsDir is deprecated [Deprecated]
```

### Additional Information
* 1.5.1 6c07b0a1f8daf96078ae68b83ead1d48675969d7
* https://github.com/nim-lang/Nim/pull/16763 (hintAsError) has the same issue

### LINKED ISSUE #20149: compiler flag `--hintAsError[XDeclaredButNotUsed]:on` causes unavoidable error in `fatal.nim` that `goToBasedException` is never used
I tried compiling code with the `--hintAsError[XDeclaredButNotUsed]:on` flag turned on.
When doing so, the compiler will **always** throw the following exception from `fatal.nim`.
This makes the flag unuseable.

### Example
```nim
# Add this to nimble file
task release, "Blub":
  --hintAsError[XDeclaredButNotUsed]:on #Can't be used because systems.nim itself has screwy stuff
  --outdir:"."
  setCommand "c", "playfile.nim"

# playfile.nim
let x = 5
echo x

# in terminal
nimble release
```

### Current Output
```
/home/philipp/.choosenim/toolchains/nim-#devel/lib/system/fatal.nim(13, 9) Error: 'gotoBasedExceptions' is declared but not used [XDeclaredButNotUsed]
```
The line this is talking about is the second one in the following code-block
```nim
when defined(nimHasExceptionsQuery):
  const gotoBasedExceptions = compileOption("exceptions", "goto") # <-- This is line 13
else:
  const gotoBasedExceptions = false
```

### Expected Output
```
5
```

### Possible Solution
I don't have a direct solution, but I see multiple ways:
1) Make it so the flag does **not** apply to foreign packages, including systems.nim! (!) This is my preferred solution (!)
2) The usage of `gotoBasedExceptions` appears to be mostly within `when` statements. Could it be that for the compiler only being used in when statements does not count to escape this check? Maybe this could be changed.

### Additional Information
I do not know when this was introduced, I only started playing around with this around nim 1.6.6 and tested again with nim devel (1.7.1) and git head (installed via `choosenim #head`).

```
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-07-17
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 0d8bec695606a65c5916d0da7fcb0a976a4e1f7b
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20149
fixes #16762

### CODE REVIEW

File: tests/errmsgs/t20149.nim [hunk 1]
```diff
@@ -0,0 +1,7 @@
+discard """
+  cmd: "nim check --hints:off --warnings:off $file"
+  joinable: false
+"""
+
+let x = 12
+echo x
```
Comment (ringabout) [inline tests/errmsgs/t20149.nim]: Before:

```
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\ansi_c.nim(54, 5) Error: 'SIGTERM' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\fatal.nim(13, 9) Error: 'gotoBasedExceptions' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\fatal.nim(55, 8) Error: 'sysFatal' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\coro_detection.nim(3, 10) Error: 'coroutinesSupportedPlatform' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\schubfach.nim(42, 6) Error: 'constructSingle' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\schubfach.nim(54, 6) Error: 'isFinite' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\schubfach.nim(57, 6) Error: 'isInf' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\schubfach.nim(61, 6) Error: 'isNaN' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\schubfach.nim(65, 6) Error: 'isZero' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\alloc.nim(935, 9) Error: 'x' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\core\locks.nim(21, 7) Error: 'insideRLocksModule' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\gc.nim(745, 9) Error: 'workPackage' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\std\syncio.nim(701, 9) Error: 'f2' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\std\syncio.nim(581, 6) Error: 'readAllFile' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\std\syncio.nim(798, 11) Error: 'insideRLocksModule' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\std\syncio.nim(95, 6) Error: 'c_feof' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(38, 8) Error: 'tryAcquireSys' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(63, 8) Error: 'signalSysCond' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(66, 8) Error: 'broadcastSysCond' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(69, 8) Error: 'initSysCond' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(71, 8) Error: 'deinitSysCond' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\syslocks.nim(73, 8) Error: 'waitSysCond' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system.nim(1784, 9) Error: 'insideRLocksModule' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system.nim(2312, 5) Error: 'SafePoint' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system.nim(267, 7) Error: 'ThisIsSystem' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system.nim(45, 3) Error: 'nil' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system.nim(552, 7) Error: 'NoFakeVars' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\alloc.nim(1006, 6) Error: 'getFreeMem' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\alloc.nim(1007, 6) Error: 'getTotalMem' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\alloc.nim(820, 6) Error: 'rawAlloc0' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\assign.nim(43, 10) Error: 'deepSeqAssignImpl' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\avltree.nim(27, 6) Error: 'find' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\cellseqs_v1.nim(18, 6) Error: 'contains' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\cellseqs_v1.nim(39, 6) Error: 'deinit' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\cellsets.nim(177, 6) Error: 'excl' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\cellsets.nim(198, 10) Error: 'elements' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\cellsets.nim(251, 10) Error: 'elementsExcept' is declared but not used [XDeclaredButNotUsed]
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\dyncalls.nim(18, 3) Error: 'NilLibHandle' is declared but not used [XDeclaredButNotUsed]
```

**Comment by ringabout** [comment]
note to future lookup: pre-existing but hintAsError/warningAsError seems not to give a correct exit code which makes testament stop to work.

**Comment by PhilippMDoerner** [comment]
In terms of feedback:
I just built this branch with koch and used it to compile my project (which is the reason I opened #20149).
It works exactly correct!

When on the branch and compiling with
`./koch temp c -d:ssl --hintAsError:XDeclaredButNotUsed --warningAsError:UnusedImport ../nimstoryfont/src/nimstoryfont.nim`
or 
`./koch temp c -d:ssl "--hintAsError[XDeclaredButNotUsed]:on" "--warningAsError[UnusedImport]:on" ../nimstoryfont/src/nimstoryfont.nim`

It no longer throws a compiler error in the system.nim or fatal.nim modules, which it previously did!

**Comment by ringabout** [comment]
Unrelated CI failure, ggplot breaks for a short while.

**Comment by PhilippMDoerner** [comment]
Given that it's running through and looking good, is this PR just waiting for somebody else to look over it?

**Comment by ringabout** [comment]
Yeah, it is ready for review. Feel free to review. Remember to approve my PR if you think it is fine.

image

File: compiler/msgs.nim [hunk 2]
```diff
@@ -531,8 +531,7 @@ proc liMessage*(conf: ConfigRef; info: TLineInfo, msg: TMsgKind, arg: string,
   of warnMin..warnMax:
     sev = Severity.Warning
     ignoreMsg = not conf.hasWarn(msg)
-    if msg in conf.warningAsErrors:
-      ignoreMsg = false
+    if not ignoreMsg and msg in conf.warningAsErrors:
```
Comment (PhilippMDoerner) [inline compiler/msgs.nim:534]: What meaning does it have for `msg in conf.warningAsErrors` to be true? What would a boolean that contained the result of that statement be accurately be named like?

Comment (ringabout) [reply compiler/msgs.nim:534]: > Maybe you could expand this test with an unused import to test that in a single go? Would only require e.g. import std/options in the file and --warningAsError:UnusedImport in trunner.nim

Sounds good to me. Feel free to add a test case in the following up prs.

> Is there a way to test that the flag correctly prohibits compilation? Or would that be out of scope?

Yeah, there are tests for that. Using compiles should work for simple cases. If you can also go through the trouble of using execCmdEx and compare the outputs. Following-up prs are welcome!

Comment (ringabout) [reply compiler/msgs.nim:534]: > What meaning does it have for msg in conf.warningAsErrors to be true?

`conf.warningAsErrors` is a set of TMsgKind enum type. If msg in the `conf.warningAsErrors`, it means this hints/warnings are enabled for the file and you have already specified that hints/warnings should be errors.

For instance, for a module within the current package,  `XDeclaredButNotUsed` is enabled. If you also specify `--hintAsError:XDeclaredButNotUsed` is true, in this case `XDeclaredButNotUsed` is the msg variable, `--hintAsError:XDeclaredButNotUsed` means `XDeclaredButNotUsed` is added to `conf.warningAsErrors`. So the statement "msg in  conf.warningAsErrors" is true.

For a foreign package, `XDeclaredButNotUsed` is not enabled by default. Even if `--hintAsError:XDeclaredButNotUsed` is specified, the statement "msg in  conf.warningAsErrors" is false.

File: tests/misc/trunner.nim [hunk 3]
```diff
@@ -249,12 +249,18 @@ sub/mmain.idx""", context
     let cmd = fmt"{nim} r -b:cpp --hints:off --nimcache:{nimcache} --warningAsError:ProveInit {file}"
     check execCmdEx(cmd) == ("witness\n", 0)
 
+  block: # bug #20149
+    let file = testsDir / "misc/m20149.nim"
+    let cmd = fmt"{nim} r --hints:off --nimcache:{nimcache} --hintAsError:XDeclaredButNotUsed {file}"
```
Comment (Varriount) [inline tests/misc/trunner.nim:254]: I'll still approve this, but would like to note that this will break if `nim` or `nimcache` are not properly escaped.

**Review by Varriount — APPROVED** [review]

**Pull request merged**